// @empirica/core/dist/admin-classic.js
import {
  Batch,
  Classic,
  ClassicListenersCollector,
  ClassicLoader,
  Context,
  ExportFormat,
  Game,
  Lobby,
  Player,
  PlayerGame,
  PlayerRound,
  PlayerStage,
  Round,
  Stage,
  classicKinds,
  evt,
  runExport,
  withTajriba
} from "./chunk-MM5L2YDU.js";
import {
  Globals as Globals2,
  TajribaAdminAccess
} from "./chunk-UKOR76YV.js";
import {
  Attributes,
  EventContext,
  ListenersCollector,
  ListenersCollectorProxy,
  Scope,
  Scopes,
  TajribaEvent
} from "./chunk-BY25FC6R.js";
import "./chunk-DGXLC63B.js";
import "./chunk-FE4Z7IXP.js";
import {
  Globals
} from "./chunk-2TT3WJZY.js";
import "./chunk-TIKLWCJI.js";
export {
  Attributes,
  Batch,
  Classic,
  ClassicListenersCollector,
  ClassicLoader,
  Context,
  EventContext,
  ExportFormat,
  Game,
  Globals2 as Globals,
  ListenersCollector,
  ListenersCollectorProxy,
  Lobby,
  Player,
  PlayerGame,
  PlayerRound,
  PlayerStage,
  Round,
  Scope,
  Scopes,
  Globals as SharedGlobals,
  Stage,
  TajribaAdminAccess,
  TajribaEvent,
  classicKinds,
  evt,
  runExport,
  withTajriba
};
